src/gululog_idx.erl

```diff
@@ -352,13 +402,132 @@ get_oldest_segid(#idx{tid = Tid}) ->
   end.
 
 %% @private Sync and and close file.
--spec file_sync_close(file:fds()) -> ok.
+-spec file_sync_close(file:fd()) -> ok.
 file_sync_close(Fd) ->
   ok = file:sync(Fd),
   ok = file:close(Fd).
 
+%% @private Delete index files.
+%% It is very important to delete files in reversed order in order to keep
+%% data integrity (in case this function crashes in the middle for example)
+%% @end
+-spec truncate_delete_do([segid()], dirname(), ?undef | dirname()) -> [file_op()].
+truncate_delete_do(DeleteSegIdList, Dir, BackupDir) ->
+  lists:map(fun(SegId) ->
+              FileName = gululog_name:mk_idx_name(Dir, SegId),
+              gululog_file:delete(FileName, BackupDir)
+            end, lists:reverse(lists:sort(DeleteSegIdList))).
+
+%% @private Truncate index file.
+-spec truncate_truncate_do(dirname(), ?undef | segid(), logid(),
+                           ?undef | dirname()) -> [file_op()].
+truncate_truncate_do(_Dir, ?undef, _LogId, _BackupDir) -> [];
+truncate_truncate_do(Dir, SegId, LogId, BackupDir) ->
+  IdxFile = gululog_name:mk_idx_name(Dir, SegId),
+  IdxPosition = get_position_in_index_file(IdxFile, LogId),
+  true = gululog_file:maybe_truncate(IdxFile, IdxPosition, BackupDir), %% assert
+  [{?OP_TRUNCATED, IdxFile}].
+
 %%%*_ TESTS ====================================================================
 
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+
+gululog_idx_test_() ->
+  [ {"truncate/5",
+      fun() ->
+         {ok, Dir}       = file:get_cwd(),
+         BackupDir       = filename:join(Dir, "backup"),
+         BackupDirDelete = filename:join(Dir, "backup_delete"),
+         [file:delete(X) || X <- gululog_name:wildcard_idx_name_reversed(Dir)
+                              ++ gululog_name:wildcard_seg_name_reversed(Dir)],
+         Idx1 = init(Dir),
+         InitLogId = 0,
+         Idx2 = append(Idx1, InitLogId + 0, 10),
+         Idx3 = append(Idx2, InitLogId + 1, 20),
+         Idx4 = append(Idx3, InitLogId + 2, 30),
+         Idx5 = append(Idx4, InitLogId + 3, 40),
+         Idx6 = append(Idx5, InitLogId + 4, 50),
+         Idx7 = switch_append(Dir, Idx6, InitLogId + 5, 60),
```

Comment:
in real case, the Position arg (last one) should always be 1 for switch_append.
60 works here in unit test, but better to make it more real.